### ISSUE TITLE
Authentication: How to know when the authentication is done after pageload?

### ISSUE THREAD

**Issue by Kamshak** [angular/angularfire2#568]
I use a custom provider to log users in via token. This works great, the user can log in and the session is persisted. When I subscribe to the auth Observable I get this:

```
1: null
2: { uid: 23123123, ... }
```

It seems that this library at first emits null, then logs the user in using the token in localStorage, then emits the authStatus. All Good. 

My Question is: How do i know when the automatic initial login has finished? If the user is simply not logged in a null is emitted, and nothing after that. Using a timeout is inconvenient and adds unnecessary delays on page load.

**Comment by hllink** [comment]
got an issue directed related with this question https://github.com/angular/angularfire2/issues/566 for redirect method of login.

**Comment by mukesh51** [comment]
@Kamshak : Timeout is inconvenient. Did you try something similar.

constructor(public auth: FirebaseAuth) {
        this.authState = auth.getAuth();
        auth.subscribe((state: FirebaseAuthState) => {
            this.authState = state;          
        });        
    }

    get authenticated(): boolean {        
        return this.authState !== null && !this.expired;
    }
    
    get expired(): boolean {
            return !this.authState || (this.authState.expires * 1000) < Date.now();
    }

**Comment by Kamshak** [comment]
Don't think it can be solved like that as there is still no way to know when to get authenticated. To clarify the problem: I want to use this in conjunction with a router so that I can redirect the user to the login page if they are not authenticated. To answer the question "Do I need to redirect the user to a login page" i need to know: "Has firebase tried to log this user in using the local session?". 

Right now all I get is this:

a) User  is logged in:
![diagram](https://hostr.co/file/R46czPJhSEDR/Unbenannt.PNG)

b) User is not logged in:
![diagram](https://hostr.co/file/pGUCrFbQBS6B/Unbenannt2.PNG)

To implement this functionality I need to know at which point "Token Checked" has happened. A simple Promise could do this as this only happens once. Or a method to try and log in using the local session.

**Comment by Toxicable** [comment]
I have also faced this issue. In the past I've used a variable like `authReady` which is awaited on by anything that wants to check to see the token.
If authReady is false then then the startup code is still running, if is true then the startup code has been run and we can safely determine if we're good to check the status of the users auth.
I find this very important for things like route guards which run on page load if the user enters the site on that page.

I think exposing an Observable or Promise (preferably Observable) with a similar concept to the above would solve this issue.
